pkg/volume/csi/csi_plugin.go

```diff
@@ -307,17 +307,32 @@ func (p *csiPlugin) GetPluginName() string {
 	return CSIPluginName
 }
 
-// GetvolumeName returns a concatenated string of CSIVolumeSource.Driver<volNameSe>CSIVolumeSource.VolumeHandle
+// GetvolumeName returns a concatenated string of DriverName<volNameSe>VolumeHandle
+// when the spec is source by a PVC and DriverName<volNameSep>volumeSpecName
 // That string value is used in Detach() to extract driver name and volumeName.
 func (p *csiPlugin) GetVolumeName(spec *volume.Spec) (string, error) {
-	csi, err := getPVSourceFromSpec(spec)
+	volSrc, pvSrc, err := getSourceFromSpec(spec)
 	if err != nil {
-		klog.Error(log("plugin.GetVolumeName failed to extract volume source from spec: %v", err))
+		klog.Error(log("plugin.GetVolumeName failed to extract volume source from spec: %s", err))
 		return "", err
 	}
 
-	// return driverName<separator>volumeHandle
-	return fmt.Sprintf("%s%s%s", csi.Driver, volNameSep, csi.VolumeHandle), nil
+	if volSrc != nil && utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
+		// When a volume is ephemeral and is sourced by a CSIVolumeSource, there
+		// is no way to properly formulate the volume name using the same scheme
+		// as when it is sourced from a PVC. A different scheme must be used to generate
+		// the volume name using a combination of the driverName and the volumeSpec as
+		//  driverName<separator>volumeSpec.Name.
+		// This is here to maintain API sanity and satisfy startup reconciliation loops.
+		// Under normal operartion, this should never be used since ephemeral volumes
+		// do not support attach/detach operations.
+		return fmt.Sprintf("%s%s%s", volSrc.Driver, volNameSep, spec.Name()), nil
```

Comment:
Should this be makeVolumeHandle not spec.Name?

Reply:
When the source is `CSIVolumeSource`, from the volume.Spec, it does not have field volumeHandle.  So, volumeHandle is not accessible at this point.  This is a workaround to how the internal VolumeAPI works as commented.

Reply:
hmm.... its not the same scheme as makeVolumeHandle uses though. What prevents a collision?

Oh. I think I see. The assertion that it shouldn't actually be used? Should it be instead set to something more obvious in that case, like a "DEADBEEF"?

Reply:
Yes, it's not used since ephemeral do not support detach which is when GetVolumeName method is invoked.  Good suggestion to set it to an obviously bad name.

Reply:
I dont think a change is suggested here.   I updated the comment to be clearer. PTAL.

Reply:
Ok taking a step back.  I think we need to carefully test out this case of multiple pods using the same volume name.

The dsw and asw key by unique volume name, and it's used for refcnting the number of users of that volume there are to determine when it's safe to unmount.

I think instead of trying to fake responses to mount device and generate potentially conflicting volume names, we should try to follow the way that in-tree ephemeral volumes actually work in Kubernetes.  

I think that may require adding a CanDeviceMount() check similar to CanAttach().  That way, when we add pods and volumes to dsw and asw, we can actually have them [do the right thing](https://github.com/kubernetes/kubernetes/blob/fccb19da9fe63dadcc52ccd81a4501fc42eedd9f/pkg/kubelet/volumemanager/cache/desired_state_of_world.go#L225)

Reply:
@msau42 I agree with the CanDeviceMount check. This would provide a clean approach that the volume manager can query.

Reply:
For now, can we agree that this is enough (exiting MountDevice if volume source is ephemeral)?

https://github.com/kubernetes/kubernetes/pull/75354/files#diff-6c5c64be2395d7a4c5088acafc82101cR300

Reply:
Actually have an issue open for this exact thing - https://github.com/kubernetes/kubernetes/issues/75352

Reply:
I think skipping the MountDevice is not too difficult.  Either add the CanDeviceMount() interface to all plugins (which is touching a lot of files), or for now, to reduce the amount of code change, we can add a special check to to `FindDeviceMountablePluginBySpec`, where it calls `CanAttach` if the volume plugin is the csi type.  Special casing csi is not great but considering that we are not accepting any more in-tree volume types, I think this is fine.

Reply:
Hm now that I think about it a little more, the second option may not work exactly. Because persistent volumes that skip attach still need to go through mount device.  So either we need to add the CanDeviceMount to the volume interface or have a special check just for CSI ephemeral volumes. @jsafrane wdyt?

Reply:
Can we fix it as it is in 1.14/alpha and properly in 1.15 (or 1.14.1)?

Reply:
the deadline for picks for 1.14.0 has passed, especially for changes oriented at an alpha feature

Reply:
@liggitt When will 1.14.1 be scheduled?

Reply:
not sure, likely 2-3ish weeks after 1.14.0 is cut, but I would not generally expect fixes for disabled-by-default alpha features to go into patch releases